typing/typeclass.ml

```diff
@@ -592,109 +596,110 @@ let rec class_field self_loc cl_num self_type meths vars arg cf =
             in
-            (val_env, met_env, par_env,Some name)
+            (class_env,Some name)
       in
-      (val_env, met_env, par_env,
+      (class_env,
        lazy (mkcf (Tcf_inherit (ovf, parent, super, inh_vars, inh_meths)))
        :: fields,
        concr_meths, warn_vals, inher, local_meths, local_vals)
 
   | Pcf_val (lab, mut, Cfk_virtual styp) ->
       if !Clflags.principal then Ctype.begin_def ();
-      let cty = Typetexp.transl_simple_type val_env false styp in
+      let cty = Typetexp.transl_simple_type class_env.val_env false styp in
       let ty = cty.ctyp_type in
       if !Clflags.principal then begin
         Ctype.end_def ();
         Ctype.generalize_structure ty
       end;
-      let (id, val_env, met_env', par_env) =
+      let (id, class_env') =
         enter_val cl_num vars false lab.txt mut Virtual ty
-          val_env met_env par_env loc
+        class_env.val_env class_env.met_env class_env.par_env loc
       in
-      (val_env, met_env', par_env,
+      (class_env',
        lazy (mkcf (Tcf_val (lab, mut, id, Tcfk_virtual cty,
-                            met_env == met_env')))
+                            class_env.met_env == class_env'.met_env)))
              :: fields,
              concr_meths, warn_vals, inher, local_meths, local_vals)
 
   | Pcf_val (lab, mut, Cfk_concrete (ovf, sexp)) ->
       if Concr.mem lab.txt local_vals then
-        raise(Error(loc, val_env, Duplicate ("instance variable", lab.txt)));
+        raise(Error(loc, class_env.val_env, Duplicate ("instance variable", lab.txt)));
       if Concr.mem lab.txt warn_vals then begin
         if ovf = Fresh then
           Location.prerr_warning lab.loc
             (Warnings.Instance_variable_override[lab.txt])
       end else begin
         if ovf = Override then
-          raise(Error(loc, val_env,
+          raise(Error(loc, class_env.val_env,
                       No_overriding ("instance variable", lab.txt)))
       end;
       if !Clflags.principal then Ctype.begin_def ();
-      let exp = type_exp val_env sexp in
+      let exp = type_exp class_env.val_env sexp in
       if !Clflags.principal then begin
         Ctype.end_def ();
         Ctype.generalize_structure exp.exp_type
        end;
-      let (id, val_env, met_env', par_env) =
+      let (id, class_env') =
         enter_val cl_num vars false lab.txt mut Concrete exp.exp_type
-          val_env met_env par_env loc
+        class_env.val_env class_env.met_env class_env.par_env loc
       in
-      (val_env, met_env', par_env,
+      (class_env',
        lazy (mkcf (Tcf_val (lab, mut, id,
-                            Tcfk_concrete (ovf, exp), met_env == met_env')))
+                            Tcfk_concrete (ovf, exp), class_env.met_env == class_env'.met_env)))
        :: fields,
        concr_meths, Concr.add lab.txt warn_vals, inher, local_meths,
        Concr.add lab.txt local_vals)
 
   | Pcf_method (lab, priv, Cfk_virtual sty) ->
```

Comment:
You could reduce the diff of the `Pcf_method` case by using a record pattern to declare three variables out of `class_env`, which avoids the need to add the record projections below.

```ocaml
let {val_env; met_env; par_env} = class_env in
(* rest of the code below, using `val_env`,etc. directly *)
```

(Note that `val_env; ` in this line is a pun for `val_env = val_env;`.)

This is a general trick to reduce the number of record projections in the code. If you update `enter_val` and `enter_meth_env` to take a `class_env` parameter, I would recommend also using it there so that the rest of the function declaration does not have to change.

I would *not* use this trick in the places where the class environment is updated in the middle of the code (with new `let class_env = ...` or `let class_env' = ...` bindings), where it is important to be explicit about which version of the class environment (the original one or an update one) we are talking about. (Over time we learn to be comfortable with this kind of shadowing and there are idioms where it is natural, but for now sticking with explicit record projections is the easiest in these scenarios.)